README.md
## AES 128 pkcs5 only support
* Provide generation of AES keys
* Provide encrypt for AES keys  
* Provide decrypt for AES keys  

* Android uses java, iOS uses OC development


## Usage

Use this package as a library
1. Depend on it
Add this to your package's pubspec.yaml file:

```
dependencies:
  flutter_aes_ecb_pkcs5: ^0.1.3
```

2. Install it You can install packages from the command line:

with Flutter:

```
$ flutter pub get
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

3. Import it Now in your Dart code, you can use:

```
import 'package:flutter_aes_ecb_pkcs5/flutter_aes_ecb_pkcs5.dart';
```



#### AES Use Example



```dart
import 'package:flutter_aes_ecb_pkcs5/flutter_aes_ecb_pkcs5.dart';

void main() {

    var data = "{\"username\":\"helloword\"}";
    
    //generate a 16-byte random key
    var key = await FlutterAesEcbPkcs5.generateDesKey(128);
    
    print(key);
    //encrypt
    var encryptText = await FlutterAesEcbPkcs5.encryptString(data, key);
    
    print(encryptText);
    //decrypt
    var decryptText  = await FlutterAesEcbPkcs5.decryptString(encryptText, key);
    
    print(decryptText);
  
}
```

#### Run Example Result
```
data:{"username":"helloword"}
create key:DA8FD2693B652D587EEE51E7ADD9DBB8
encryptText :851544E224F6C1C7F560AB82889DE0DF66EEB6762AB836F4EE13E31100F0EA16
decryptText :{"username":"helloword"}
```

#### License

The MIT License(http://opensource.org/licenses/MIT)

Please feel free to use and contribute to the development.




CHANGELOG.md
## 0.0.1

* first create aes ecb pkcs5 .

## 0.0.2

* update readme and Code comment.

## 0.1.0
* update readme and Code comment and description.
* its release.

## 0.1.1
* change description.

## 0.1.2
* update Flutter SDK is 2.x.
* support null safety
* update android gradle version 4.x

## 0.1.3
* update Flutter SDK is 3.x.
* support null safety
* update android gradle version 7.x

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_aes_ecb_pkcs5_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_aes_ecb_pkcs5_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {

  const MethodChannel channel = MethodChannel('flutter_aes_ecb_pkcs5');

  setUp(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger.setMockMethodCallHandler(
      channel,
          (MethodCall methodCall) async {
        return '42';
      },
    );
  });

  tearDown(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger.setMockMethodCallHandler(channel, null);
  });


}


lib/flutter_aes_ecb_pkcs5.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// AES ECB PKCS5 encrypt & decrypt.
class FlutterAesEcbPkcs5 {

  static const MethodChannel _channel =
      const MethodChannel('flutter_aes_ecb_pkcs5');

  /// Get device system version.
  ///
  /// * If it is a Android device, like 'Android 9.0'.
  /// * If it is a IOS device, like 'IOS 11.2'.
  static Future<String?> get platformVersion async {
    final String? version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  /// Used to generate a specific length of AES key with the given [length].
  ///
  /// If it is a 128-bit key, enter 128 and the returned key will be encoded by Hex.
  /// [value] and [key] shoudn't be null.
  static Future<String?> generateDesKey(int length) async {
    try {
      final String? result =
          await _channel.invokeMethod('generateDesKey', {"length": length});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }

  /// Encrypts the [txt] with the given [key].
  ///
  /// The result of the encryption will be encoded by HEX.
  /// [txt] and [key] shoudn't be null.
  static Future<String?> encryptString(String txt, String key) async {
    try {
      final String? result =
          await _channel.invokeMethod('encrypt', {"input": txt, "key": key});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }

  /// Decrypts the [txt] with the given [key].
  ///
  /// The content [txt] that needs to be decrypted must be HEX encoded，key also.
  /// [txt] and [key] shoudn't be null.
  static Future<String?> decryptString(String txt, String key) async {
    try {
      final String? result =
          await _channel.invokeMethod('decrypt', {"input": txt, "key": key});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }
}


ios/Classes/FlutterAesEcbPkcs5Plugin.m
#import "FlutterAesEcbPkcs5Plugin.h"
#import <CommonCrypto/CommonCryptor.h>
#import <Foundation/Foundation.h>


@implementation FlutterAesEcbPkcs5Plugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_aes_ecb_pkcs5"
            binaryMessenger:[registrar messenger]];
  FlutterAesEcbPkcs5Plugin* instance = [[FlutterAesEcbPkcs5Plugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {

      NSDictionary *argsMap  = call.arguments;

    if ([@"getPlatformVersion" isEqualToString:call.method]) {
      result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    } else if ([@"generateDesKey" isEqualToString:call.method]) {

        id length = [argsMap objectForKey:@"length"];
        NSLog(@"obj = %@", length);

        NSString *key = [FlutterAesEcbPkcs5Plugin randomlyGeneratedBitString:length];
        //NSLog(@"AES加密的秘钥key=%@",key);

        NSString *hexKey = [FlutterAesEcbPkcs5Plugin hexStringFromString:key];

        //NSLog(@"AES加密的秘钥hexKey=%@",hexKey);

        result(hexKey.uppercaseString);

    }else if ([@"encrypt" isEqualToString:call.method]) {

        id input = [argsMap objectForKey:@"input"];
        //NSLog(@"obj = %@", input);
        id key = [argsMap objectForKey:@"key"];
        //NSLog(@"obj = %@", key);

        NSString *encode = [FlutterAesEcbPkcs5Plugin encyptPKCS5:input WithKey:key];

        //NSLog(@"结果=%@",encode.uppercaseString);
        result(encode.uppercaseString);
    } else if ([@"decrypt" isEqualToString:call.method]) {

        id input = [argsMap objectForKey:@"input"];
        //NSLog(@"obj = %@", input);
        id key = [argsMap objectForKey:@"key"];
        //NSLog(@"obj = %@", key);

        NSString *encode = [FlutterAesEcbPkcs5Plugin decrypPKCS5:input WithKey:key];

        //NSLog(@"结果=%@",encode);
        result(encode);

    } else {
      result(FlutterMethodNotImplemented);
    }

}


+ (NSString *)randomlyGeneratedBitString:(id)length
{
    NSString *string = [[NSString alloc]init];

    NSLog(@"len->%@",length);
    int len = [length intValue] / 8 ;
    NSLog(@"len->%d",len);

    for (int i = 0; i < len; i++) {
        int number = arc4random() % 36;
        if (number < 10) {
            int figure = arc4random() % 10;
            NSString *tempString = [NSString stringWithFormat:@"%d", figure];
            string = [string stringByAppendingString:tempString];
        }else {
            int figure = (arc4random() % 26) + 97;
            char character = figure;
            NSString *tempString = [NSString stringWithFormat:@"%c", character];
            string = [string stringByAppendingString:tempString];
        }
    }
    return  string;
}


//字符串加密(16进制)
+ (NSString *)encyptPKCS5:(NSString *)plainText WithKey:(NSString *)key{

    //把string 转NSData
    NSData* data = [plainText dataUsingEncoding:NSUTF8StringEncoding];

    //length
    size_t plainTextBufferSize = [data length];

    const void *vplainText = (const void *)[data bytes];


    uint8_t *bufferPtr = NULL;
    size_t bufferPtrSize = 0;
    size_t movedBytes = 0;

    bufferPtrSize = (plainTextBufferSize + kCCBlockSizeAES128) & ~(kCCBlockSizeAES128 - 1);
    bufferPtr = malloc( bufferPtrSize * sizeof(uint8_t));
    memset((void *)bufferPtr, 0x0, bufferPtrSize);
    //解密hex加密的密钥
    char *publicKey = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:key];

    NSLog(@"vkey->len->%d",strlen(publicKey));

    //配置CCCrypt
    CCCryptorStatus ccStatus = CCCrypt(kCCEncrypt,
                                       kCCAlgorithmAES128, //3DES
                                       kCCOptionECBMode|kCCOptionPKCS7Padding, //设置模式
                                       publicKey,    //key
                                       kCCKeySizeAES128,
                                       nil,     //偏移量，这里不用，设置为nil;不用的话，必须为nil,不可以为@“”
                                       vplainText,
                                       plainTextBufferSize,
                                       (void *)bufferPtr,
                                       bufferPtrSize,
                                       &movedBytes);

    if (ccStatus == kCCSuccess) {
        NSData *myData = [NSData dataWithBytes:(const char *)bufferPtr length:(NSUInteger)movedBytes];

        //16进制(你也可以换成base64等)
        NSUInteger          len = [myData length];
        char *              chars = (char *)[myData bytes];
        NSMutableString *   hexString = [[NSMutableString alloc] init];

        for(NSUInteger i = 0; i < len; i++ )
            [hexString appendString:[NSString stringWithFormat:@"%0.2hhx", chars[i]]];
        return hexString;
    }

    free(bufferPtr);
    return nil;

}


//解密
+(NSString *)decrypPKCS5:(NSString *)encryptText WithKey:(NSString *)key
{

    //解密hex加密的密钥
    char *publicKey = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:key];

//    NSLog(@"publicKey->publicKey->%s",publicKey);
//    NSLog(@"vkey->len->%d",strlen(publicKey));

    char *data1 = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:encryptText];

    int plainTextBufferSize = [encryptText length] / 2 ;

    //NSLog(@"data1->plainTextBufferSize->%d",plainTextBufferSize);

    NSData *content=[NSData dataWithBytes:data1 length:plainTextBufferSize];

    //NSLog(@"data1->content->%@",content);

    NSUInteger dataLength = [content length];
    size_t bufferSize = dataLength + kCCBlockSizeAES128;
    void *buffer = malloc(bufferSize);

    size_t numBytesCrypted = 0;
    CCCryptorStatus cryptStatus = CCCrypt(kCCDecrypt,
                                          kCCAlgorithmAES128,
                                          kCCOptionPKCS7Padding|kCCOptionECBMode,
                                          publicKey,
                                          kCCBlockSizeAES128,
                                          NULL,
                                          [content bytes],
                                          dataLength,
                                          buffer,
                                          bufferSize,
                                          &numBytesCrypted);

    if (cryptStatus == kCCSuccess) {
        NSData *resultData = [NSData dataWithBytesNoCopy:buffer length:numBytesCrypted];

        return [[NSString alloc] initWithData:resultData encoding:NSUTF8StringEncoding];
    }
    free(buffer);
    return nil;
}

+ (char *)convertHexStrToChar:(NSString *)hexString {

    int mallocLen = [hexString length] / 2 + 1;

    char *myBuffer = (unsigned char *)malloc(mallocLen);

    memset(myBuffer,'\0',mallocLen);

    for (int i = 0; i < [hexString length] - 1; i += 2) {
        unsigned int anInt;
        NSString * hexCharStr = [hexString substringWithRange:NSMakeRange(i, 2)];
        NSScanner * scanner = [[NSScanner alloc] initWithString:hexCharStr];
        [scanner scanHexInt:&anInt];
        myBuffer[i / 2] = (char)anInt;
    }
    return myBuffer;
}


+ (NSString *)hexStringFromString:(NSString *)string{
    NSData *myD = [string dataUsingEncoding:NSUTF8StringEncoding];
    Byte *bytes = (Byte *)[myD bytes];
    //下面是Byte 转换为16进制。
    NSString *hexStr=@"";
    for(int i=0;i<[myD length];i++)
    {
        NSString *newHexStr = [NSString stringWithFormat:@"%x",bytes[i]&0xff];///16进制数
        if([newHexStr length]==1)
            hexStr = [NSString stringWithFormat:@"%@0%@",hexStr,newHexStr];
        else
            hexStr = [NSString stringWithFormat:@"%@%@",hexStr,newHexStr];
    }
    return hexStr;
}


@end


android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/Hex.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import java.util.Locale;

public class Hex{

	/**
	 * 将字节数组进行HEX编码.
	 * @param data 待编码的字节数组
	 * @return 编码后的HEX字符串
	 */
	public static String encode(byte[] data) {

		StringBuilder sb = new StringBuilder();
		String stmp;
		for (int n = 0; n < data.length; n++){
			stmp = (Integer.toHexString(data[n] & 0xFF));
			if (stmp.length() == 1)
				sb.append("0");
			sb.append(stmp);
		}
		return sb.toString().toUpperCase(Locale.CHINA);
	}

	/**
	 * 将HEX编码的字符串转换为字节数组.
	 * @param data 待转换的HEX字符串
	 * @return 转换后的字节数组
	 */
	public static byte[] decode(String data) {

		int n = data.length();
		if (n % 2 != 0){
			data = "0" + data;
			n++;
		}
		n = (n / 2);
		byte[] bytes = new byte[n];
		for (int i = 0; i < n; i++){
			bytes[i] = (byte)Integer
					.parseInt(data.substring(i * 2, i * 2 + 2), 16);
		}
		return bytes;
	}

}


android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/AesSecurity.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import java.security.NoSuchAlgorithmException;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

/**
 * aes加密算法
 * AES/ECB/PKCS5Padding
 */
public class AesSecurity {

	private static final String ALGORITHM = "AES";
	private static final String TRANSFORMATION = "AES/ECB/PKCS5Padding";
	private static final String DEFUALT_ENCODING = "UTF8";


	public  String generateDesKey(int length)  {
		//实例化
		KeyGenerator kgen = null;
		try {
			kgen = KeyGenerator.getInstance("AES");
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		//设置密钥长度
		kgen.init(length);
		//生成密钥
		SecretKey skey = kgen.generateKey();
		//返回密钥的二进制编码
		return Hex.encode(skey.getEncoded());
	}


	/**
	 * 加密
     * @param  input 加密的字符串
     * @param  key   解密的key ,十六进制表示
	 * @return HexString
     */
	 
	public String encrypt(String input, String key){

		byte[] decodeKey = Hex.decode(key);

		byte[] crypted = null;
		try{
			SecretKeySpec skey = new SecretKeySpec(decodeKey, ALGORITHM);
			Cipher cipher = Cipher.getInstance(TRANSFORMATION);
			cipher.init(Cipher.ENCRYPT_MODE, skey);
			crypted = cipher.doFinal(input.getBytes());
		}catch(Exception e){
			System.out.println(e.toString());
		}
		return parseByte2HexStr(crypted);

	}

	/**
	 * 将二进制转换成16进制
     * @param buf
     * @return
     */
    private static String parseByte2HexStr(byte buf[]) {
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < buf.length; i++) {
            String hex = Integer.toHexString(buf[i] & 0xFF);
            if (hex.length() == 1) {
                hex = '0' + hex;
            }
            sb.append(hex.toUpperCase());
        }
        return sb.toString();
    }
	
	/**
	 * 将16进制转换为二进制
     * @param hexStr
     * @return
     */
    private static byte[] parseHexStr2Byte(String hexStr) {
		
        if (hexStr.length() < 1)
            return null;
        byte[] result = new byte[hexStr.length()/2];
        for (int i = 0;i< hexStr.length()/2; i++) {
            int high = Integer.parseInt(hexStr.substring(i*2, i*2+1), 16);
            int low = Integer.parseInt(hexStr.substring(i*2+1, i*2+2), 16);
            result[i] = (byte) (high * 16 + low);
        }
        return result;
		
    }
	
	/**
	 * 解密
     * @param  input 解密的字符串
     * @param  key   解密的key
	 * @return String
     */
	public String decrypt(String input, String key){
		byte[] decodeKey = Hex.decode(key);
		byte[] output = null;
		try{
			SecretKeySpec skey = new SecretKeySpec(decodeKey, ALGORITHM);
			Cipher cipher = Cipher.getInstance(TRANSFORMATION);
			cipher.init(Cipher.DECRYPT_MODE, skey);
			output = cipher.doFinal(parseHexStr2Byte(input));
		}catch(Exception e){
			System.out.println(e.toString());
		}
		return new String(output);
	}
}


android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/FlutterAesEcbPkcs5Plugin.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** FlutterAesEcbPkcs5Plugin */
public class FlutterAesEcbPkcs5Plugin implements FlutterPlugin, MethodCallHandler  {

  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    channel = new MethodChannel(binding.getBinaryMessenger(), "flutter_aes_ecb_pkcs5");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }else if (call.method.equals("generateDesKey")) {
      int length = call.argument("length");
      if (length != 0 ) {
        try {
          String encoded = new AesSecurity().generateDesKey(length);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "generateDesKey failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "generateDesKey failure.", null);
      }

    }else if(call.method.equals("encrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String encoded = new AesSecurity().encrypt(text, publicKey);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Encrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Encrypt failure.", null);
      }
    }else if(call.method.equals("decrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String decoded = new AesSecurity().decrypt(text, publicKey);
          result.success(decoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Decrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Decrypt failure.", null);
      }
    } else {
      result.notImplemented();
    }
  }



  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_aes_ecb_pkcs5/flutter_aes_ecb_pkcs5.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String pkcs5Progress;

    //string format
    var data = "{\"username\":\"helloword\"}";

    //create 16 byte random key
    var key = await FlutterAesEcbPkcs5.generateDesKey(128);

    print(key);

    //encrypt
    var encryptText = await FlutterAesEcbPkcs5.encryptString(data,key!);

    print(encryptText);
    //decrypt
    var decryptText  = await FlutterAesEcbPkcs5.decryptString(encryptText!, key);

    print(decryptText);

    pkcs5Progress = "data:"+data +"\n"+
        "create key:"+key +"\n"
        +"encryptText :"+encryptText +"\n"
        +"decryptText :"+decryptText! +"\n";

    print(pkcs5Progress);
    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = pkcs5Progress;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('flutter_aes_ecb_pkcs5'),
        ),
        body: Center(
          child: Text('pkcs5Progress:\n $_platformVersion\n'),
        ),
      ),
    );
  }
}
